File: integrations/javascript/rik-client.js
/**
 * RIK JavaScript/Node.js Client
 * ==============================
 *
 * Official JavaScript client library for the Recursive Intelligence Kernel (RIK).
 * Works in both Node.js and browser environments.
 *
 * Example (Node.js):
 *   const RIKClient = require('./rik-client');
 *   const client = new RIKClient('http://localhost:8000');
 *   const result = await client.processInvoice(pdfContent, 'INV-001');
 *   console.log(result.final_action);
 *
 * Example (Browser):
 *   <script src="rik-client.js"></script>
 *   <script>
 *     const client = new RIKClient('http://localhost:8000');
 *     client.processInvoice(pdfContent, 'INV-001')
 *       .then(result => console.log(result.final_action));
 *   </script>
 */

class RIKClient {
    /**
     * Create a new RIK client.
     *
     * @param {string} baseUrl - Base URL of RIK API (e.g., 'http://localhost:8000')
     * @param {Object} options - Client options
     * @param {string} options.apiKey - Optional API key for authentication
     * @param {number} options.timeout - Request timeout in milliseconds (default: 30000)
     * @param {number} options.maxRetries - Maximum number of retries (default: 3)
     */
    constructor(baseUrl = 'http://localhost:8000', options = {}) {
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
        this.apiKey = options.apiKey || null;
        this.timeout = options.timeout || 30000;
        this.maxRetries = options.maxRetries || 3;

        // Detect environment
        this.isNode = typeof process !== 'undefined' && process.versions && process.versions.node;

        // Use node-fetch in Node.js environment
        if (this.isNode && typeof fetch === 'undefined') {
            try {
                this.fetch = require('node-fetch');
            } catch (e) {
                console.warn('node-fetch not found. Please install it: npm install node-fetch');
                this.fetch = null;
            }
        } else {
            this.fetch = fetch;
        }
    }

    /**
     * Make HTTP request with retry logic and error handling.
     *
     * @private
     */
    async _request(method, endpoint, data = null, params = null, retryCount = 0) {
        const url = new URL(`${this.baseUrl}${endpoint}`);

        // Add query parameters
        if (params) {
            Object.keys(params).forEach(key =>
                url.searchParams.append(key, params[key])
            );
        }

        // Build headers
        const headers = {
            'Content-Type': 'application/json',
        };

        if (this.apiKey) {
            headers['X-API-Key'] = this.apiKey;
        }

        // Build request options
        const options = {
            method: method,
            headers: headers,
        };

        if (data) {
            options.body = JSON.stringify(data);
        }

        try {
            // Create timeout promise
            const timeoutPromise = new Promise((_, reject) =>
                setTimeout(() => reject(new Error('Request timeout')), this.timeout)
            );

            // Make request with timeout
            const response = await Promise.race([
                this.fetch(url.toString(), options),
                timeoutPromise
            ]);

            // Handle HTTP errors
            if (response.status === 401 || response.status === 403) {
                throw new RIKAuthenticationError('Authentication failed. Check your API key.');
            }

            if (response.status === 422) {
                const errorData = await response.json();
                throw new RIKValidationError(errorData.detail || 'Validation error', errorData.errors);
            }

            if (response.status === 429) {
                const retryAfter = response.headers.get('Retry-After');
                throw new RIKRateLimitError(retryAfter);
            }

            if (response.status >= 400) {
                let errorMessage = `API request failed with status ${response.status}`;
                try {
                    const errorData = await response.json();
                    errorMessage = errorData.detail || errorMessage;
                } catch (e) {
                    // Ignore JSON parse error
                }
                throw new RIKAPIError(errorMessage, response.status, endpoint);
            }

            // Parse successful response
            const responseData = await response.json();
            return responseData;

        } catch (error) {
            // Retry on network errors
            if (retryCount < this.maxRetries &&
                (error.message === 'Request timeout' || error.name === 'FetchError')) {

                const delay = Math.pow(2, retryCount) * 1000; // Exponential backoff
                await new Promise(resolve => setTimeout(resolve, delay));
                return this._request(method, endpoint, data, params, retryCount + 1);
            }

            // Re-throw RIK errors
            if (error instanceof RIKError) {
                throw error;
            }

            // Wrap other errors
            if (error.message === 'Request timeout') {
                throw new RIKTimeoutError(this.timeout / 1000, endpoint);
            }

            if (error.message.includes('fetch')) {
                throw new RIKConnectionError(url.toString(), error);
            }

            throw new RIKAPIError(error.message, 0, endpoint);
        }
    }

    // ========================================================================
    // CORE ENDPOINTS
    // ========================================================================

    /**
     * Execute a reasoning task.
     *
     * @param {string} task - Task description to process
     * @returns {Promise<Object>} Task result with reasoning output
     */
    async runTask(task) {
        return await this._request('POST', '/run_task', { task });
    }

    /**
     * Process invoice with intelligent exception handling.
     *
     * @param {string} pdfContent - PDF content as string or JSON
     * @param {string} invoiceId - Optional invoice identifier
     * @param {Object} context - Optional additional context
     * @returns {Promise<Object>} Invoice processing result with decision and reasoning
     */
    async processInvoice(pdfContent, invoiceId = null, context = null) {
        const payload = {
            pdf_content: pdfContent,
            invoice_id: invoiceId,
            context: context
        };
        return await this._request('POST', '/process_invoice', payload);
    }

    /**
     * Recover broken web scraper selector.
     *
     * @param {string} failedSelector - The selector that stopped working
     * @param {string} html - Current HTML content
     * @param {string} url - URL being scraped
     * @param {Object} context - Optional context about what should be selected
     * @returns {Promise<Object>} Selector recovery result with new selector and reasoning
     */
    async recoverSelector(failedSelector, html, url, context = null) {
        const payload = {
            failed_selector: failedSelector,
            html: html,
            url: url,
            context: context
        };
        return await this._request('POST', '/recover_selector', payload);
    }

    /**
     * Test if a selector works on given HTML.
     *
     * @param {string} selector - CSS or XPath selector to test
     * @param {string} html - HTML content to test against
     * @param {string} selectorType - 'css' or 'xpath' (default: 'css')
     * @returns {Promise<Object>} Test result with success status and matched elements
     */
    async testSelector(selector, html, selectorType = 'css') {
        const payload = {
            selector: selector,
            html: html,
            selector_type: selectorType
        };
        return await this._request('POST', '/test_selector', payload);
    }

    // ========================================================================
    // MONITORING & ANALYTICS
    // ========================================================================

    /**
     * Get API health status.
     *
     * @returns {Promise<Object>} Health status with subsystem checks
     */
    async getHealth() {
        return await this._request('GET', '/health');
    }

    /**
     * Check if API is alive (simple liveness check).
     *
     * @returns {Promise<boolean>} True if API is responding
     */
    async isAlive() {
        try {
            const result = await this._request('GET', '/health/live');
            return result.status === 'alive';
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if API is ready to handle requests.
     *
     * @returns {Promise<boolean>} True if API is ready
     */
    async isReady() {
        try {
            const result = await this._request('GET', '/health/ready');
            return result.status === 'ready';
        } catch (error) {
            return false;
        }
    }

    /**
     * Get performance and efficiency metrics.
     *
     * @returns {Promise<Object>} Metrics with efficiency stats
     */
    async getMetrics() {
        return await this._request('GET', '/metrics');
    }

    /**
     * Get invoice processing statistics and ROI data.
     *
     * @returns {Promise<Object>} Invoice stats with automation rates and savings
     */
    async getInvoiceStats() {
        return await this._request('GET', '/invoice_stats');
    }

    /**
     * Get API version information.
     *
     * @returns {Promise<Object>} Version and environment info
     */
    async getVersion() {
        return await this._request('GET', '/version');
    }

    /**
     * Retrieve episodic memory entries.
     *
     * @param {number} limit - Maximum number of entries to return (default: 10)
     * @returns {Promise<Array>} List of memory episodes
     */
    async getMemory(limit = 10) {
        const result = await this._request('GET', '/memory', null, { limit });
        return result.episodes || [];
    }
}

// ============================================================================
// EXCEPTION CLASSES
// ============================================================================

class RIKError extends Error {
    constructor(message, details = {}) {
        super(message);
        this.name = 'RIKError';
        this.details = details;
    }
}

class RIKConnectionError extends RIKError {
    constructor(url, originalError = null) {
        super(`Failed to connect to RIK API at ${url}`);
        this.name = 'RIKConnectionError';
        this.url = url;
        this.originalError = originalError;
    }
}

class RIKAPIError extends RIKError {
    constructor(message, statusCode, endpoint = null) {
        super(message);
        this.name = 'RIKAPIError';
        this.statusCode = statusCode;
        this.endpoint = endpoint;
    }
}

class RIKAuthenticationError extends RIKAPIError {
    constructor(message = 'Authentication failed. Check your API key.') {
        super(message, 401);
        this.name = 'RIKAuthenticationError';
    }
}

class RIKValidationError extends RIKAPIError {
    constructor(message, validationErrors = null) {
        super(`Validation error: ${message}`, 422);
        this.name = 'RIKValidationError';
        this.validationErrors = validationErrors;
    }
}

class RIKTimeoutError extends RIKError {
    constructor(timeoutSeconds, endpoint = null) {
        super(`Request timed out after ${timeoutSeconds}s`);
        this.name = 'RIKTimeoutError';
        this.timeoutSeconds = timeoutSeconds;
        this.endpoint = endpoint;
    }
}

class RIKRateLimitError extends RIKAPIError {
    constructor(retryAfter = null) {
        let message = 'Rate limit exceeded';
        if (retryAfter) {
            message += `. Retry after ${retryAfter} seconds`;
        }
        super(message, 429);
        this.name = 'RIKRateLimitError';
        this.retryAfter = retryAfter;
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RIKClient;
    module.exports.RIKError = RIKError;
    module.exports.RIKConnectionError = RIKConnectionError;
    module.exports.RIKAPIError = RIKAPIError;
    module.exports.RIKAuthenticationError = RIKAuthenticationError;
    module.exports.RIKValidationError = RIKValidationError;
    module.exports.RIKTimeoutError = RIKTimeoutError;
    module.exports.RIKRateLimitError = RIKRateLimitError;
}

File: integrations/csharp/RIKClient.cs
/**
 * RIK C#/.NET Client
 * ===================
 *
 * Official C# client library for the Recursive Intelligence Kernel (RIK).
 * Compatible with .NET Framework 4.5+ and .NET Core/5+
 * Perfect for UiPath integration.
 *
 * Example:
 *   var client = new RIKClient("http://localhost:8000");
 *   var result = await client.ProcessInvoiceAsync(pdfContent, "INV-001");
 *   Console.WriteLine(result.FinalAction);
 */

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace RIK
{
    /// <summary>
    /// Client for interacting with the Recursive Intelligence Kernel (RIK) API.
    /// </summary>
    public class RIKClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;
        private readonly string _apiKey;

        /// <summary>
        /// Initializes a new instance of the RIKClient class.
        /// </summary>
        /// <param name="baseUrl">Base URL of the RIK API (e.g., "http://localhost:8000")</param>
        /// <param name="apiKey">Optional API key for authentication</param>
        /// <param name="timeoutSeconds">Request timeout in seconds (default: 30)</param>
        public RIKClient(string baseUrl, string apiKey = null, int timeoutSeconds = 30)
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _apiKey = apiKey;

            _httpClient = new HttpClient
            {
                Timeout = TimeSpan.FromSeconds(timeoutSeconds)
            };

            _httpClient.DefaultRequestHeaders.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/json")
            );

            if (!string.IsNullOrEmpty(_apiKey))
            {
                _httpClient.DefaultRequestHeaders.Add("X-API-Key", _apiKey);
            }
        }

        // ====================================================================
        // CORE ENDPOINTS
        // ====================================================================

        /// <summary>
        /// Execute a reasoning task.
        /// </summary>
        public async Task<TaskResult> RunTaskAsync(string task)
        {
            var payload = new { task = task };
            var response = await PostAsync<JObject>("/run_task", payload);
            return new TaskResult(response);
        }

        /// <summary>
        /// Process invoice with intelligent exception handling.
        /// </summary>
        public async Task<InvoiceProcessingResult> ProcessInvoiceAsync(
            string pdfContent,
            string invoiceId = null,
            Dictionary<string, object> context = null)
        {
            var payload = new
            {
                pdf_content = pdfContent,
                invoice_id = invoiceId,
                context = context
            };

            var response = await PostAsync<JObject>("/process_invoice", payload);
            return new InvoiceProcessingResult(response);
        }

        /// <summary>
        /// Recover broken web scraper selector.
        /// </summary>
        public async Task<SelectorRecoveryResult> RecoverSelectorAsync(
            string failedSelector,
            string html,
            string url,
            Dictionary<string, object> context = null)
        {
            var payload = new
            {
                failed_selector = failedSelector,
                html = html,
                url = url,
                context = context
            };

            var response = await PostAsync<JObject>("/recover_selector", payload);
            return new SelectorRecoveryResult(response);
        }

        /// <summary>
        /// Test if a selector works on given HTML.
        /// </summary>
        public async Task<SelectorTestResult> TestSelectorAsync(
            string selector,
            string html,
            string selectorType = "css")
        {
            var payload = new
            {
                selector = selector,
                html = html,
                selector_type = selectorType
            };

            var response = await PostAsync<JObject>("/test_selector", payload);
            return new SelectorTestResult(response);
        }

        // ====================================================================
        // MONITORING & ANALYTICS
        // ====================================================================

        /// <summary>
        /// Get API health status.
        /// </summary>
        public async Task<HealthStatus> GetHealthAsync()
        {
            var response = await GetAsync<JObject>("/health");
            return new HealthStatus(response);
        }

        /// <summary>
        /// Check if API is alive (simple liveness check).
        /// </summary>
        public async Task<bool> IsAliveAsync()
        {
            try
            {
                var response = await GetAsync<JObject>("/health/live");
                return response["status"]?.ToString() == "alive";
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Check if API is ready to handle requests.
        /// </summary>
        public async Task<bool> IsReadyAsync()
        {
            try
            {
                var response = await GetAsync<JObject>("/health/ready");
                return response["status"]?.ToString() == "ready";
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Get performance and efficiency metrics.
        /// </summary>
        public async Task<MetricsResponse> GetMetricsAsync()
        {
            var response = await GetAsync<JObject>("/metrics");
            return new MetricsResponse(response);
        }

        /// <summary>
        /// Get invoice processing statistics and ROI data.
        /// </summary>
        public async Task<InvoiceStats> GetInvoiceStatsAsync()
        {
            var response = await GetAsync<JObject>("/invoice_stats");
            return new InvoiceStats(response);
        }

        /// <summary>
        /// Get API version information.
        /// </summary>
        public async Task<VersionInfo> GetVersionAsync()
        {
            var response = await GetAsync<JObject>("/version");
            return new VersionInfo(response);
        }

        /// <summary>
        /// Retrieve episodic memory entries.
        /// </summary>
        public async Task<List<JObject>> GetMemoryAsync(int limit = 10)
        {
            var response = await GetAsync<JObject>($"/memory?limit={limit}");
            var episodes = response["episodes"]?.ToObject<List<JObject>>();
            return episodes ?? new List<JObject>();
        }

        // ====================================================================
        // HTTP REQUEST HELPERS
        // ====================================================================

        private async Task<T> GetAsync<T>(string endpoint)
        {
            var url = $"{_baseUrl}{endpoint}";
            var response = await _httpClient.GetAsync(url);
            await EnsureSuccessStatusCodeAsync(response);

            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<T>(content);
        }

        private async Task<T> PostAsync<T>(string endpoint, object data)
        {
            var url = $"{_baseUrl}{endpoint}";
            var json = JsonConvert.SerializeObject(data);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _httpClient.PostAsync(url, content);
            await EnsureSuccessStatusCodeAsync(response);

            var responseContent = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<T>(responseContent);
        }

        private async Task EnsureSuccessStatusCodeAsync(HttpResponseMessage response)
        {
            if (response.IsSuccessStatusCode)
                return;

            var errorContent = await response.Content.ReadAsStringAsync();
            var statusCode = (int)response.StatusCode;

            switch (statusCode)
            {
                case 401:
                case 403:
                    throw new RIKAuthenticationException("Authentication failed. Check your API key.");

                case 422:
                    throw new RIKValidationException($"Validation error: {errorContent}");

                case 429:
                    throw new RIKRateLimitException("Rate limit exceeded");

                default:
                    throw new RIKAPIException($"API request failed with status {statusCode}: {errorContent}");
            }
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }

    // ========================================================================
    // RESPONSE MODELS
    // ========================================================================

    public class InvoiceProcessingResult
    {
        public string InvoiceId { get; set; }
        public string FinalAction { get; set; }
        public double ConfidenceScore { get; set; }
        public string Reasoning { get; set; }
        public int ExceptionsFound { get; set; }
        public int ExceptionsResolved { get; set; }
        public double ProcessingTimeSeconds { get; set; }
        public bool TraditionalRPAWouldFail { get; set; }
        public int SimilarCasesFound { get; set; }
        public int StrategiesSimulated { get; set; }

        public InvoiceProcessingResult(JObject data)
        {
            InvoiceId = data["invoice_id"]?.ToString();
            FinalAction = data["final_action"]?.ToString();
            ConfidenceScore = data["confidence_score"]?.ToObject<double>() ?? 0.0;
            Reasoning = data["reasoning"]?.ToString();
            ExceptionsFound = data["exceptions_found"]?.ToObject<int>() ?? 0;
            ExceptionsResolved = data["exceptions_resolved"]?.ToObject<int>() ?? 0;
            ProcessingTimeSeconds = data["processing_time_seconds"]?.ToObject<double>() ?? 0.0;
            TraditionalRPAWouldFail = data["traditional_rpa_would_fail"]?.ToObject<bool>() ?? false;
            SimilarCasesFound = data["similar_cases_found"]?.ToObject<int>() ?? 0;
            StrategiesSimulated = data["strategies_simulated"]?.ToObject<int>() ?? 0;
        }

        public bool NeedsHumanReview()
        {
            return FinalAction == "escalate" || ConfidenceScore < 0.7;
        }
    }

    public class SelectorRecoveryResult
    {
        public string RecoveredSelector { get; set; }
        public string SelectorType { get; set; }
        public double ConfidenceScore { get; set; }
        public string Reasoning { get; set; }
        public double ProcessingTimeSeconds { get; set; }
        public List<string> FallbackStrategies { get; set; }

        public SelectorRecoveryResult(JObject data)
        {
            RecoveredSelector = data["recovered_selector"]?.ToString();
            SelectorType = data["selector_type"]?.ToString();
            ConfidenceScore = data["confidence_score"]?.ToObject<double>() ?? 0.0;
            Reasoning = data["reasoning"]?.ToString();
            ProcessingTimeSeconds = data["processing_time_seconds"]?.ToObject<double>() ?? 0.0;
            FallbackStrategies = data["fallback_strategies"]?.ToObject<List<string>>() ?? new List<string>();
        }
    }

    public class SelectorTestResult
    {
        public bool Success { get; set; }
        public int ElementCount { get; set; }
        public string SampleText { get; set; }

        public SelectorTestResult(JObject data)
        {
            Success = data["success"]?.ToObject<bool>() ?? false;
            ElementCount = data["element_count"]?.ToObject<int>() ?? 0;
            SampleText = data["sample_text"]?.ToString();
        }
    }

    public class TaskResult
    {
        public string Task { get; set; }
        public string Result { get; set; }
        public string Reasoning { get; set; }
        public double ProcessingTimeSeconds { get; set; }

        public TaskResult(JObject data)
        {
            Task = data["task"]?.ToString();
            Result = data["result"]?.ToString();
            Reasoning = data["reasoning"]?.ToString();
            ProcessingTimeSeconds = data["processing_time_seconds"]?.ToObject<double>() ?? 0.0;
        }
    }

    public class HealthStatus
    {
        public string Status { get; set; }
        public Dictionary<string, bool> Subsystems { get; set; }

        public HealthStatus(JObject data)
        {
            Status = data["status"]?.ToString();
            Subsystems = data["subsystems"]?.ToObject<Dictionary<string, bool>>() ?? new Dictionary<string, bool>();
        }

        public bool IsHealthy()
        {
            return Status == "healthy" || Status == "alive";
        }
    }

    public class MetricsResponse
    {
        public double Efficiency { get; set; }
        public int TotalEpisodes { get; set; }
        public int SuccessfulEpisodes { get; set; }
        public int FailedEpisodes { get; set; }

        public MetricsResponse(JObject data)
        {
            var metrics = data["metrics"] ?? data;
            Efficiency = metrics["efficiency"]?.ToObject<double>() ?? 0.0;
            TotalEpisodes = metrics["total_episodes"]?.ToObject<int>() ?? 0;
            SuccessfulEpisodes = metrics["successful_episodes"]?.ToObject<int>() ?? 0;
            FailedEpisodes = metrics["failed_episodes"]?.ToObject<int>() ?? 0;
        }
    }

    public class InvoiceStats
    {
        public int TotalInvoicesProcessed { get; set; }
        public int InvoicesWithExceptions { get; set; }
        public int ExceptionsAutoResolved { get; set; }
        public int ExceptionsEscalated { get; set; }
        public double AutomationRate { get; set; }
        public double TraditionalRPAAutomationRate { get; set; }
        public double AnnualSavingsUSD { get; set; }
        public string AutomationImprovement { get; set; }

        public InvoiceStats(JObject data)
        {
            var stats = data["stats"] ?? data;
            TotalInvoicesProcessed = stats["total_invoices_processed"]?.ToObject<int>() ?? 0;
            InvoicesWithExceptions = stats["invoices_with_exceptions"]?.ToObject<int>() ?? 0;
            ExceptionsAutoResolved = stats["exceptions_auto_resolved"]?.ToObject<int>() ?? 0;
            ExceptionsEscalated = stats["exceptions_escalated"]?.ToObject<int>() ?? 0;
            AutomationRate = stats["automation_rate"]?.ToObject<double>() ?? 0.0;
            TraditionalRPAAutomationRate = stats["traditional_rpa_automation_rate"]?.ToObject<double>() ?? 0.0;
            AnnualSavingsUSD = stats["annual_savings_usd"]?.ToObject<double>() ?? 0.0;
            AutomationImprovement = stats["automation_improvement"]?.ToString();
        }
    }

    public class VersionInfo
    {
        public string Version { get; set; }
        public string Environment { get; set; }

        public VersionInfo(JObject data)
        {
            Version = data["version"]?.ToString();
            Environment = data["environment"]?.ToString();
        }
    }

    // ========================================================================
    // EXCEPTION CLASSES
    // ========================================================================

    public class RIKException : Exception
    {
        public RIKException(string message) : base(message) { }
        public RIKException(string message, Exception innerException) : base(message, innerException) { }
    }

    public class RIKAPIException : RIKException
    {
        public RIKAPIException(string message) : base(message) { }
    }

    public class RIKAuthenticationException : RIKAPIException
    {
        public RIKAuthenticationException(string message) : base(message) { }
    }

    public class RIKValidationException : RIKAPIException
    {
        public RIKValidationException(string message) : base(message) { }
    }

    public class RIKRateLimitException : RIKAPIException
    {
        public RIKRateLimitException(string message) : base(message) { }
    }
}
